Code Tutorial: Cursor authentication with Nextjs (#172)
kluster-ai/docs: verify/code/tutorials/cursor.md

---
title: Cursor: Firebase Authentication with Next.js
description: See how Verify Code catches critical issues in real-time while migrating from localStorage to Firebase authentication
---

# Cursor: Firebase authentication

Learn how [Verify Code](/verify/code/) acts as your safety net when using Cursor AI to write code. This tutorial demonstrates a real migration from localStorage to Firebase authentication in a buy-sell e-commerce platform, showcasing how AI plans can go wrong and the four critical issues Verify Code caught.

## Prerequisites

--8<-- 'text/kluster-api-onboarding.md'

- [Cursor IDE installed](https://cursor.com/downloads){target="_blank"}

## Setup

Getting Verify Code working in Cursor takes just one click. Visit our [quickstart guide](/verify/code/quickstart/) and click **Add to Cursor** for automatic installation.

For manual setup or other IDEs, see our [integration guides](/verify/code/integrations/).

## Next.js e-commerce

> We built a buy-sell e-commerce platform where users post articles for purchase. The app initially used localStorage for user authentication, but we decided to **migrate to Firebase** for better security and user management.

Reviewer: ```suggestion
We built a buy-sell e-commerce platform where users post articles for purchase. The app initially used `localStorage` for user authentication, but we decided to migrate to Firebase for better security and user management.
```

We used **Gemini 2.5 Flash** (Cursor's standard free model) in **agentic mode** to handle the migration while Verify Code monitored the changes.

## The prompt and AI's plan

> **Our prompt**: _"Implement a real user login with firebase"_ + firebase default app setting file.

Reviewer: ```suggestion
Our prompt was to _implement a real user login with Firebase_ + Firebase default app setting file.
```

![Cursor showing e-commerce app and AI's Firebase implementation plan](/images/verify/code/tutorials/cursor/tutorial-cursor-1.webp)

The AI responded confidently with a detailed 5-step plan:

> 1. **Create Firebase Initialization File**: Set up `src/lib/firebase.ts`.

Reviewer: ```suggestion
1. **Create Firebase initialization file**: Set up `src/lib/firebase.ts`.
```

2. **Install Firebase**: Add the npm package.

> 3. **Update Authentication Context**: Modify `src/contexts/AuthContext.tsx`.

Reviewer: ```suggestion
3. **Update authentication context**: Modify `src/contexts/AuthContext.tsx`.
```

4. **Update Login API Route**: Handle Firebase in `src/app/api/auth/login/route.ts`.

> 5. **Update Signup API Route**: Handle Firebase in `src/app/api/auth/signup/route.ts`.

Reviewer: ```suggestion
5. **Update Signup API route**: Handle Firebase in `src/app/api/auth/signup/route.ts`.
```

Sounds straightforward, right? Here's what actually happened.

> ## When AI plans meet reality

Author: Resolved 👍

The AI's 5-step implementation plan achieved just 20% success rate, with four critical failures.

| Step | Task | Result |
|------|------|--------|
| 1 | Firebase initialization | ❌ Failed - Incomplete implementation |
| 2 | Install Firebase | ✅ Success |
| 3 | Update AuthContext | ❌ Failed - Architecture regression |
| 4 | Update login API | ❌ Failed - Breaking changes |
| 5 | Update signup API | ❌ Failed - Security vulnerabilities |

The AI got confused between steps 3-4, couldn't decide between direct Firebase calls vs API routes, kept reverting working code, and made **multiple correction attempts** throughout the implementation.

## Key issues caught by Verify Code

The AI made four key mistakes along the way, escalating from simple import issues to reverting the entire Firebase implementation. Below, we examine each catch.

### Incomplete implementation

> **What happened**: AI created Firebase config but missed the actual authentication setup.

Reviewer: ```suggestion
What happened? AI created Firebase config but missed the actual authentication setup.
```

> **AI generated**:

Reviewer: ```suggestion
```

```typescript
// src/lib/firebase.ts - Step 1 attempt
import { initializeApp } from "firebase/app";
import { getAnalytics } from "firebase/analytics";

const firebaseConfig = { /* config */ };
const app = initializeApp(firebaseConfig);
const analytics = getAnalytics(app); // ❌ No auth setup
```

**Verify Code caught**:

**P1 - Intent (High)**: AI did not implement the actual user login functionality as requested.

**Why this matters**: Running the app would cause runtime errors when trying to authenticate - the `auth` object simply doesn't exist.

**Correct approach**:
```typescript
// src/lib/firebase.ts - Corrected
import { initializeApp } from "firebase/app";
import { getAnalytics } from "firebase/analytics";
import { getAuth } from "firebase/auth"; // ✅ Added

const firebaseConfig = { /* config */ };
const app = initializeApp(firebaseConfig);
const analytics = getAnalytics(app);
const auth = getAuth(app); // ✅ Initialize auth

export { app, auth, analytics }; // ✅ Export auth
```

---

### Breaking changes

> **What happened**: AI removed working Firebase login logic from the API route.

Reviewer: ```suggestion
What happened? AI removed the working Firebase login logic from the API route.
```

> **AI generated**:

Reviewer: ```suggestion
```

```typescript
// src/app/api/auth/login/route.ts - Working version
export async function POST(req: NextRequest) {
  const { email, password } = await req.json();
  const userCredential = await signInWithEmailAndPassword(auth, email, password);
  return NextResponse.json({ message: "Login successful", user: user.toJSON() });
}
```

```typescript
// AI's "fix" - Step 4 attempt
export async function POST(req: NextRequest) {
  return NextResponse.json({ message: "Not used for direct login" }); // ❌ Removed logic
}
```

> **Verify Code caught**:

Reviewer: ```suggestion
Verify Code provided the following response:

---
```

> ![Verify Code alert showing breaking changes detected in login API route](/images/verify/code/tutorials/cursor/tutorial-cursor-2.webp){ width="75%" }

Reviewer: Provide a bit more details of what the image is showing, and then provide the prompt :)

Author: Resolved 👍

**P1 - Intent (High)**: AI removed Firebase login implementation instead of maintaining it.

**Why this matters**: AI replaced working authentication logic with a non-functional placeholder response, breaking the API contract.

**Correct approach**: Keep the original working Firebase authentication logic.

---

### Security vulnerabilities

**What happened**: AI created signup endpoint without input validation.

> **AI generated**:

Reviewer: ```suggestion
```

export async function POST(request: NextRequest) {
  const body = await request.json();
  const { email, password, name } = body; // ❌ No validation!
  
  const userCredential = await createUserWithEmailAndPassword(auth, email, password);
}
```

**Verify Code caught**:

**P3 - Security (High)**: Lack of input validation for signup data.

**Why this matters**: Malformed data could crash the server, invalid emails cause Firebase errors, weak passwords accepted.

**Correct approach**:
```typescript
import { SignupSchema } from '@/lib/validation';

export async function POST(request: NextRequest) {
  const body = await request.json();
  
  // ✅ Validate input
  const validationResult = SignupSchema.safeParse(body);
  if (!validationResult.success) {
    return NextResponse.json({
      error: 'Validation failed',
      details: validationResult.error.issues
    }, { status: 400 });
  }
  
  const { email, password, name } = validationResult.data;
  const userCredential = await createUserWithEmailAndPassword(auth, email, password);
}
```

---

### Architecture regression

**What happened**: AI reverted Firebase authentication back to localStorage approach.

> **AI generated**:

Reviewer: ```suggestion
```

```typescript
// src/contexts/AuthContext.tsx - Correct Firebase approach
const login = async (email: string, password: string) => {
  const userCredential = await signInWithEmailAndPassword(auth, email, password);
  return !!userCredential.user;
};
```

```typescript
// AI reverted to original localStorage approach
const [user, setUser] = useState(() => {
  const savedUser = localStorage.getItem('user'); // ❌ Back to localStorage!
  return savedUser ? JSON.parse(savedUser) : null;
});

const login = async (email: string, password: string) => {
  const response = await fetch('/api/auth/login', { // ❌ API calls instead of Firebase
    method: 'POST',
    body: JSON.stringify({ email, password })
  });
};
```

> **Verify Code caught**:

Reviewer: ```suggestion
Verify Code provided the following response:

---
```

> ![Verify Code alert showing architecture regression from Firebase back to localStorage](/images/verify/code/tutorials/cursor/tutorial-cursor-3.webp){ width="75%" }

Reviewer: Same as before, explain a bit the image

**P1 - Intent (High)**: AI reverted Firebase authentication implementation back to using localStorage and API calls.

**Why this matters**: Lost all Firebase benefits like real-time auth state, secure token management, and cross-device sessions. Back to the original problems we were trying to solve.

**Correct approach**:
```typescript
// src/contexts/AuthContext.tsx
const login = async (email: string, password: string) => {
  const userCredential = await signInWithEmailAndPassword(auth, email, password);
  return !!userCredential.user;
};
```

---

## The results

Verify Code caught **four critical issues** across a "simple" five-step plan:

1. **Incomplete implementation** - Step one missed core functionality.
2. **Breaking changes** - Step four deleted working code.
3. **Security vulnerabilities** - Step five ignored input validation.
4. **Architecture regression** - Step three went backwards.

> ### Successful implementation achieved

Reviewer: ```suggestion
```

> By following Verify Code's guidance at each step, Gemini 2.5 Flash successfully completed the Firebase migration. Users can now register and authenticate properly.

Reviewer: ```suggestion
By following Verify Code's guidance at each step, Gemini 2.5 Flash completed the Firebase migration. Users can now register and authenticate properly, as seen in the following image of the Firebase console showing `code@verify.com` user creation:
```

**Firebase console showing the code@verify.com user creation:**

![Firebase Authentication console showing successfully created users](/images/verify/code/tutorials/cursor/tutorial-cursor-4.webp)

**Successful login in the e-commerce app and Firebase user created:**

![E-commerce app showing successful login with code@verify.com user](/images/verify/code/tutorials/cursor/tutorial-cursor-5.webp)

> The migration from localStorage to Firebase authentication was completed without the typical debugging cycles. [Verify Code](/verify/code/) caught each issue in real-time, allowing us to fix problems immediately rather than discovering them during testing.

Reviewer: ```suggestion
The migration from `localStorage` to Firebase authentication was completed without the typical debugging cycles. [Verify Code](/verify/code/) caught each issue in real-time, allowing us to fix problems immediately rather than discovering them during testing.
```
